// metaform/lib/js/components/grid/paginator.js
var React = require('react');
window.$ = window.jQuery = require('jquery');

var Paginator = React.createClass({displayName: "Paginator",

  getInitialState: function () {
    return {pageSize: this.props.pageSize};
  },

  componentWillReceiveProps: function (nextProps) {
    this.setState({pageSize: nextProps.pageSize});
  },

  handleChange: function (ev) {
    this.setState({pageSize: ev.target.value});
  },

  componentDidMount: function () {
    $(this.refs.pageSizeSelector.getDOMNode()).on('changed.fu.combobox', function (ev, opt) {
      this.props.onPageSizeChange(opt.value || opt.text);
    }.bind(this));
  },

  first: function (event) {
    event.preventDefault();
    this.props.onPageChange('first');
  },

  previous: function (event) {
    event.preventDefault();
    this.props.onPageChange('prev');
  },

  next: function (event) {
    event.preventDefault();
    this.props.onPageChange('next');
  },

  last: function (event) {
    event.preventDefault();
    this.props.onPageChange('last');
  },

  render: function () {
    var firstClassNames = 'first';
    var prevClassNames = 'prev';
    var lastClassNames = 'last';
    var nextClassNames = 'next';
    var page = this.props.page;
    var pageSize = this.props.pageSize;
    var totalEntries = this.props.totalEntries;
    var totalPages = parseInt(totalEntries / pageSize, 10) || 1;
    if (totalPages * pageSize < totalEntries) {
      totalPages += 1;
    }
    if (page === 1) {
      firstClassNames += ' disabled';
      prevClassNames += ' disabled';
    }
    if (page === totalPages) {
      lastClassNames += ' disabled';
      nextClassNames += ' disabled';
    }
    var start = (page - 1) * pageSize + 1;
    var end = start + Math.min(pageSize, totalEntries) - 1;

    return (
      React.createElement("ul", {className: "pagination pagination-small"}, 
        React.createElement("li", {className: firstClassNames}, 
          React.createElement("a", {href: "#", onClick: this.first}, "First")
        ), 
        React.createElement("li", {className: prevClassNames}, 
          React.createElement("a", {href: "#", onClick: this.previous}, "« Previous")
        ), 
        React.createElement("li", {className: "current"}, 
          React.createElement("div", {className: "page-size-selector"}, 
            React.createElement("span", null, start, " to ", end, " of ", this.props.totalEntries, " | items/page:"), 
            React.createElement("div", {className: "input-group input-append dropdown combobox", "data-initialize": "combobox", ref: "pageSizeSelector"}, 
              React.createElement("input", {type: "text", className: "form-control", value: this.state.pageSize, onChange: this.handleChange}), 
              React.createElement("div", {className: "input-group-btn"}, 
                React.createElement("button", {type: "button", className: "btn btn-default dropdown-toggle", "data-toggle": "dropdown"}, React.createElement("span", {className: "caret"})), 
                React.createElement("ul", {className: "dropdown-menu dropdown-menu-right"}, 
                  React.createElement("li", {"data-value": "10"}, React.createElement("a", {href: "#"}, "10")), 
                  React.createElement("li", {"data-value": "50"}, React.createElement("a", {href: "#"}, "50")), 
                  React.createElement("li", {"data-value": "100"}, React.createElement("a", {href: "#"}, "100")), 
                  React.createElement("li", {"data-value": "500"}, React.createElement("a", {href: "#"}, "500"))
                )
              )
            )
          )
        ), 
        React.createElement("li", {className: nextClassNames}, 
          React.createElement("a", {href: "#", onClick: this.next}, "Next »")
        ), 
        React.createElement("li", {className: lastClassNames}, 
          React.createElement("a", {href: "#", onClick: this.last}, "Last")
        )
      )
    );
  }
});

module.exports = Paginator;
