// metaform/lib/js/components/form/confirm-button.js
var React = require('react');
window.$ = window.jQuery = require('jquery');

var ConfirmButton = React.createClass({displayName: "ConfirmButton",

  componentDidMount: function () {
    var self = this;
    $('[data-toggle="popover"]').popover()
        .on('shown.bs.popover', function () {
          $('.btn-cancel').click(self.cancel);
          $('.btn-ok').click(self.ok);
        })
        .on('hidden.bs.popover', function () {
          $('.btn-cancel').off();
          $('.btn-ok').off();
        });
  },

  componentWillUnmount: function () {
    $('[data-toggle="popover"]').popover('destroy');
  },

  cancel: function () {
    $('[data-toggle="popover"]').popover('hide');
  },

  ok: function () {
    if (typeof(this.props.onAccept) === 'function') {
      this.props.onAccept();
    }
    $('[data-toggle="popover"]').popover('hide');
  },

  render: function () {
    var markup = (
      React.createElement("div", {className: "confirm-popup"}, 
        React.createElement("p", null, "This action cannot be undone."), 
        React.createElement("div", {className: "text-center"}, 
          React.createElement("div", {className: "btn-group", role: "group", "aria-label": "..."}, 
            React.createElement("button", {type: "button", className: "btn btn-default btn-sm btn-cancel"}, 
              React.createElement("span", {className: "fa fa-times-circle"}), 
              "Cancel"
            ), 
            React.createElement("button", {type: "button", className: "btn btn-primary btn-sm btn-ok"}, 
              React.createElement("span", {className: "fa fa-check-circle-o"}), 
              "OK"
            )
          )
        )
      )
    );
    return (
      React.createElement("button", {type: "button", className: "btn btn-danger", "data-toggle": "popover", 
              title: "Please confirm", "data-placement": "left", "data-container": "body", 
              "data-content": React.renderToString(markup), "data-html": true}, this.props.title)
    );
  }
});

module.exports = ConfirmButton;
