// metaform/lib/js/components/form/form.js
var React = require('react');
var AppActions = require('../../actions/app-actions');
var FormStore = require('../../stores/form-store');
var FormStoreWatchMixin = require('../../mixins/form-store-watch-mixin');
var serialize = require('form-serialize');
var bootstrap = require('bootstrap');
var fuelux = require('fuelux');
var _ = require('../../helpers/functional');
window.$ = window.jQuery = require('jquery');

var ds1url = window.apiBaseURL || '/api';

var valueFunctions = [];

// TODO
// assumes there is only one custom metadata section
var hasCustom = false;

var Form = React.createClass({displayName: "Form",

  mixins: [FormStoreWatchMixin(getFormSchema)],

  componentWillMount: function () {
    AppActions.getFormSchema(this.props.name);
  },

  // componentDidMount: function () {
  //   $('.checkbox').checkbox();
  // },

  handleSubmit: function (event) {
    event.preventDefault();
    var form = this.refs[this.props.name + '-form'].getDOMNode();
    //var str = serialize(form, {hash: true}); // url encoded
    var serialized = serialize(form, {hash: true});

    var obj = {};
    var key;
    var properties;
    var spec = this.state.schema;
    var schema = spec.schema;
    if ('properties' in schema) {
      properties = schema.properties;
    } else {
      properties = schema;
    }

    // initialize the object to update the server model for empty form fields
    for (key in properties) {
      if (properties[key].type === 'array' &&
          key in spec.form &&
          'type' in spec.form[key] &&
          spec.form[key].type === 'multiselect')
      {
        obj[key] = [];
      } else {
        obj[key] = null;
      }
    }

    // convert booleans
    for (key in serialized) {
      var val = serialized[key];
      if (val === 'on') {
        obj[key] = true;
      } else {
        obj[key] = val;
      }
    }
    for (var i = 0; i < valueFunctions.length; i++) {
      //obj = Object.assign({}, obj, valueFunctions[i](obj));
      obj = $.extend({}, obj, valueFunctions[i](obj));
    }
    if (hasCustom) {
      var customFields = this.refs.custom.getFields();
      var temp = {};
      for (i = 0; i < customFields.length; i++) {
        key = customFields[i];
        temp[key] = obj[key];
        delete obj[key];
      }
      obj['custom_metadata'] = temp;
    }

    var save = true;
    if (typeof(this.props.onPreSubmit) === 'function') {
      save = this.props.onPreSubmit(obj);
    }

    if (save) {
      var item = this.props.value;
      if ('_links' in item) {
        AppActions.updateEntity(item._links.self.href, obj);
      } else {
        AppActions.createEntity(this.props.name, obj);
      }
      if (typeof(this.props.onSubmitted) === 'function') {
        this.props.onSubmitted(obj);
      }
    }
  },

  render: function () {
    var schema = this.state.schema;
    var formType = this.props.formType || 'horizontal';
    var value = this.props.value || {};
    var fields = createElements(schema, formType, value);
    var submitButton = (
      React.createElement("button", {ref: "submitBtn", type: "button", className: "btn btn-default", 
              onClick: this.handleSubmit, "data-loading-text": "Updating...", 
              autoComplete: "off"}, "Save")
    );
    if (formType === 'horizontal') {
      return (
        React.createElement("form", {id: this.props.name + '-form', ref: this.props.name + '-form', role: "form", className: "form-horizontal"}, 
          fields, 
          React.createElement("div", {className: "form-group"}, 
            React.createElement("div", {className: "col-sm-offset-2 col-sm-10"}, 
              submitButton
            )
          )
        )
      );
    } else {
      return (
        React.createElement("form", {id: this.props.name + '-form', ref: this.props.name + '-form', role: "form"}, 
          fields, 
          submitButton
        )
      );
    }
  }
});

function getFormSchema() {
  return {schema: FormStore.getFormSchema()};
}

function createElements(schema, formType, value) {
  var properties;
  var i;
  if ('schema' in schema) {
    if ('properties' in schema.schema) {
      properties = schema.schema.properties;
    } else {
      properties = schema.schema;
    }

    var formOptions = schema.form;

    // TODO
    // The form options array was intended to communicate field order
    // as JSON object keys are unordered, as per the spec.
    // Enhance to use form options index to order fields, if given.
    if (formOptions && formOptions.constructor === Array) {
      var opts = {};
      for (i = 0; i < formOptions.length; i++) {
        var opt = formOptions[i];
        var key = opt.key;
        delete opt.key;
        opts[key] = opt;
      }
      formOptions = opts;
    }

    var tabbedLayout = formOptions.tabbedLayout;
    if (tabbedLayout) {
      var tabs = [], tab;
      var panes = [], pane;
      var li, fields, fieldname;
      for (i = 0; i < tabbedLayout.length; i++) {
        tab = tabbedLayout[i];
        fields = [];
        for (var j = 0; j < tab.fields.length; j++) {
          fieldname = tab.fields[j];
          fields.push(createField(formType, value, formOptions, fieldname, properties[fieldname]));
        }
        if (i == 0) {
          li = (
            React.createElement("li", {role: "presentation", className: "active"}, 
              React.createElement("a", {href: '#' + tab.id, "aria-controls": tab.id, role: "tab", "data-toggle": "tab"}, tab.title)
            )
          );
          pane = (
            React.createElement("div", {role: "tabpanel", className: "tab-pane active", id: tab.id}, 
              fields
            )
          );
        } else {
          li = (
            React.createElement("li", {role: "presentation"}, 
              React.createElement("a", {href: '#' + tab.id, "aria-controls": tab.id, role: "tab", "data-toggle": "tab"}, tab.title)
            )
          );
          pane = (
            React.createElement("div", {role: "tabpanel", className: "tab-pane", id: tab.id}, 
              fields
            )
          );
        }
        tabs.push(li);
        panes.push(pane);
      }
      return (
        React.createElement("div", null, 
          React.createElement("ul", {className: "nav nav-tabs", role: "tablist"}, 
            tabs
          ), 
          React.createElement("div", {className: "tab-content"}, 
            panes
          )
        )
      );
    } else {
      return _.mapObject(function (k, v) {
        return createField(formType, value, formOptions, k, v);
      }, properties);
    }
  } else {
    return createLabel('Loading...');
  }
}

function createField(formType, value, formOptions, k, v) {
  var opts = $.extend({}, {key: k, field: k, value: value[k], formType: formType}, v);
  if (formOptions && (k in formOptions)) {
    opts = $.extend({}, opts, formOptions[k]);
  }
  if (v.type === 'string' && !('enum' in v)) {
    if (opts.escape) {
      opts.value = encodeURI(opts.value);
    }
    if (opts.type === 'textarea') {
      return createTextarea(opts);
    }
    return createTextInput(opts);

  } else if (v.type === 'integer') {
    return createTextInput(opts);

  } else if (v.type === 'array' || 'enum' in v) {
    if (opts.type === 'pillbox') {
      return createPillbox(opts);
    }
    if (opts.type === 'repeater') {
      return createRepeater(opts);
    }
    if (!v.hasOwnProperty('enum')) {
      opts.value = value[k + 'Id'];
    }
    return createSelect(opts);

  } else if (v.type === 'boolean') {
    return createBool(opts);

  } else if (v.type === 'object') {
    return createCustom(opts);
  }
}

function createLabel(text) {
  return React.createElement('label', null, text);
}

function createTextInput(opts) {
  var TextInput = require('./text-input');
  return React.createElement(TextInput, opts);
}

function createTextarea(opts) {
  var Textarea = require('./textarea');
  return React.createElement(Textarea, opts);
}

function createPillbox(opts) {
  var field = opts.field;
  var Pillbox = require('./pillbox');
  valueFunctions.push(function () {
    var obj = {};
    obj[field] = $('#' + field).pillbox('items').map(function (obj) {
        return obj.text;
      });
    return obj;
  });
  return React.createElement(Pillbox, opts);
}

function createBool(opts) {
  var Bool = require('./bool');
  return React.createElement(Bool, opts);
}

function createSelect(opts) {
  var field = opts.field;
  var Select = require('./select');
  if (opts.type === 'multiselect') {
    opts.multi = true;
    valueFunctions.push(function (data) {
      var obj = {};
      if (data && data[field]) {
        if (typeof data[field] === 'string') {
          obj[field] = data[field].split(',');
        } else {
          obj[field] = data[field];
        }
      } else {
        obj[field] = null;
      }
      return obj;
    });
  } else {
    opts.multi = false;
  }
  return React.createElement(Select, opts);
}

// TODO
function createRepeater(opts) {
  var Repeater = require('./repeater');
  valueFunctions.push(function () {
    return {
      customerIdMappings: [
        {
          pk: {
            customerIdType: ds1url + '/customer-id-types/662',
            eventType: ds1url + '/event-types/42'
          },
          customerIdExpression: "#this['Customer_ID_Siebel']"
        }
      ]
    };
  });
  return React.createElement(Repeater, opts);
}

function createCustom(opts) {
  var Custom = require('./custom');
  // TODO
  // assumes there is only one custom metadata section
  opts['ref'] = 'custom';
  hasCustom = true;
  return React.createElement(Custom, opts);
}

module.exports = Form;
